packages/langium/src/workspace/documents.ts

```diff
@@ -282,10 +306,28 @@ export interface LangiumDocuments {
     addDocument(document: LangiumDocument): void;
 
     /**
-     * Retrieve the document with the given URI, if present. Otherwise create a new document
-     * and add it to the managed documents.
+     * Retrieve the document with the given URI, if present. Otherwise returns `undefined`.
+     */
+    getDocument(uri: URI): LangiumDocument | undefined;
+
+    /**
+     * Retrieve the document with the given URI. If not present, a new one will be created using the file system access.
+     * The new document will be added to the list of documents managed under this service.
+     */
+    getOrCreateDocument(uri: URI, cancellationToken?: CancellationToken): Promise<LangiumDocument>;
+
+    /**
+     * Creates a new document with the given URI and text content.
+     * The new document is automatically added to this service and can be retrieved using {@link getDocument}.
+     */
+    createDocument(uri: URI, text: string): LangiumDocument;
+
+    /**
+     * Creates a new document with the given URI and text content asynchronously.
+     * The process can be interrupted with a cancellation token.
+     * The new document is automatically added to this service and can be retrieved using {@link getDocument}.
      */
-    getOrCreateDocument(uri: URI): LangiumDocument;
```

Comment:
is there a reason not to keep the old api.
writing

`hasDocument ? getDocument : createDocument` might be a bit tedious

Reply:
That would require something like this:

```ts
getOrCreateDocument(uri: URI, text: string | () => string): LangiumDocument;
```

Would that be fine for you? I aim to deprecate the synchronous `readFileSync` method of the file system access.

Reply:
hmmm why do i need to pass a string? when i tried to call hasDocument i get falses in our tests in 2.1.3
=> who put the documents into the map?

Reply:
Maybe You've overseen the new `getOrCreateDocument(uri: URI, cancellationToken?: CancellationToken): Promise<LangiumDocument>` method? The method still exists, it's just async now. The signature above would be necessary to keep it sync, because we cannot load from disk in a sync way anymore.